File: Leetcode 49 - group-anagrams/49.group-anagrams.cpp
/*
 * @lc app=leetcode id=49 lang=cpp
 *
 * [49] Group Anagrams
 */

// @lc code=start
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  vector<vector<string>> groupAnagrams(vector<string> &strs) {
    vector<vector<string>> res;
    unordered_map<string, vector<string>> check;

    for (string s : strs) {
      string sorted = s;
      sort(s.begin(), s.end());
      check[s].push_back(sorted);
    }

    for (const auto &pair : check) {
      res.push_back(pair.second);
    }

    return res;
  }
};

int main() {
  Solution sol;
  vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
  auto res = sol.groupAnagrams(strs);
  for (auto &vec : res) {
    for (auto &s : vec) {
      cout << s << ", ";
    }
  }

  cout << endl;
  vector<string> strs2 = {""};
  auto res2 = sol.groupAnagrams(strs2);
  for (auto &vec : res2) {
    for (auto &s : vec) {
      cout << s << " ";
    }
    cout << endl;
  }
}
// @lc code=end

File: Leetcode 49 - group-anagrams/49.group-anagrams.java
/*
 * @lc app=leetcode id=49 lang=java
 *
 * [49] Group Anagrams
 */

// @lc code=start

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

class Solution {

  public List<List<String>> groupAnagrams(String[] strs) {
    HashMap<String, List<String>> map = new HashMap<>();

    // Sort each string and use it as a key in the map
    for (String s : strs) {
      // Sort the string
      char[] chars = s.toCharArray();
      Arrays.sort(chars);
      String sorted = new String(chars);

      // If map doesn't have the key, add it
      // If map has the key, add the string to the list
      map.computeIfAbsent(sorted, k -> new ArrayList<>()).add(s);
    }

    return new ArrayList<>(map.values());
  }

  public static void main(String[] args) {
    Solution sol = new Solution();

    System.out.println(
      sol.groupAnagrams(
        new String[] { "eat", "tea", "tan", "ate", "nat", "bat" }
      )
    );
    System.out.println(sol.groupAnagrams(new String[] { "" }));
    System.out.println(sol.groupAnagrams(new String[] { "a" }));
  }
}
// @lc code=end
